### PR TITLE
8316594: C2 SuperWord: wrong result with hand unrolled loops

### PR DESCRIPTION
**Pull request by eme64**
Please read description in https://github.com/openjdk/jdk/pull/15864.

Currently, we only compare the first element of the load-pack with the StoreVector. This is not correct, we need to compare all elements of the load-pack with the StoreVector. Otherwise it may be that the first element lays below/before the memory region of the StoreVector (Less), but the last element of the load-pack for example could have overlap with the StoreVector (Equal).

Tier1-6 + stress-testing.
Performance testing.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8316594](https://bugs.openjdk.org/browse/JDK-8316594): C2 SuperWord: wrong result with hand unrolled loops (**Bug** - P3)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [1b6b0582](https://git.openjdk.org/jdk/pull/15866/files/1b6b05827b41de32744f296fd6a65b5b9ed5c7d9)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15866/head:pull/15866` \
`$ git checkout pull/15866`

Update a local copy of the PR: \
`$ git checkout pull/15866` \
`$ git pull https://git.openjdk.org/jdk.git pull/15866/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15866`

View PR using the GUI difftool: \
`$ git pr show -t 15866`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15866.diff">https://git.openjdk.org/jdk/pull/15866.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15866#issuecomment-1731397647)

### CODE REVIEW

**Comment by vnkozlov** [comment]
You still need to add description to this PR what you are changing.

**Review by vnkozlov — COMMENTED** [review]
I have the same point as in [15864](https://github.com/openjdk/jdk/pull/15864) - wht if we know that they are different based on where they come from.
I think in both PRs tests should have such cases - combination of local and external arrays.

**Comment by vnkozlov** [comment]
> @vnkozlov I suggest we add those tests when we actually implement the detection of separate arrays. At that point, we can also create appropriate IR rules.
> Besides, I think the tests would be testing something somewhat different. The bug here happens because we have two method parameters which both reference to the same array, but we do not know that at compile time.
> Your suggested tests would exercise all sorts of combinations of different arrays on different parameters, if I understand correctly?

Not only on different parameters but also locally allocated, loaded from fields (instance and static), returned by a called method. In short, when EA may give you answer that arrays are different.

Anyway, as we discussed in #15864 it will be separate RFE. And current change is conservative - possible less vectorizations when arrays are different. But I am fine with that.

**Review by vnkozlov — APPROVED** [review]
Good.

File: test/hotspot/jtreg/compiler/loopopts/superword/TestMovingLoadBeforeStore.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+
+/**
+ * @test
+ * @requires vm.compiler2.enabled
+ * @bug 8316679
+ * @summary In SuperWord::output, LoadVector can be moved before StoreVector, but only if it is proven to be safe.
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/loopopts/superword/TestMovingLoadBeforeStore.java:29]: The test should have `@key randomness`

File: test/hotspot/jtreg/compiler/loopopts/superword/TestMovingLoadBeforeStore.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    static void test_ref(byte[] a, byte[] b, int inv) {
+        for (int i = 0; i < RANGE-4; i+=4) {
+            a[i +  0]++;
+            a[i +  1]++;
+            a[i +  2]++;
+            a[i +  3]++;
+            b[inv + i +  0]++;
+            b[inv + i +  1]++;
+            b[inv + i +  2]++;
+            b[inv + i +  3]++;
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/loopopts/superword/TestMovingLoadBeforeStore.java:96]: You might want to remove the extra whitespaces in the index expressions.

File: test/hotspot/jtreg/compiler/loopopts/superword/TestMovingLoadBeforeStore2.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+ *                   -XX:LoopUnrollLimit=250 -Xbatch
+ *                   -XX:+UnlockDiagnosticVMOptions -XX:+StressLCM
+ *                   compiler.loopopts.superword.TestMovingLoadBeforeStore2
+ */
+
+package compiler.loopopts.superword;
+import java.util.Random;
+import jdk.test.lib.Utils;
+
+public class TestMovingLoadBeforeStore2 {
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/loopopts/superword/TestMovingLoadBeforeStore2.java:41]: I think the two tests should be merged or is there a reason to have them separate? If you want different command line arguments, you can add two `@run`. Even if they are not needed for both, I think it doesn't hurt.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me. To avoid merge conflicts when pushing https://github.com/openjdk/jdk/pull/15864, you could fix both issues in the same PR since they are in the same code.

**Comment by eme64** [comment]
@TobiHartmann I want to push them as separate fixes, because one can be back-ported, but not the other one:
> We could fix both issues at once. But I am fixing them separately, first JDK-8316679, which can be backported to JDK-11, and then JDK-8316594 which only seems to have caused regressions in JDK-22.

I'll unify the tests once the other PR is integrated.

**Comment by eme64** [comment]
@TobiHartmann @vnkozlov merged with master after https://github.com/openjdk/jdk/pull/15864.
And added one more test with unsafe (test3).
I did not change the VM code, only the tests. Quick re-review would be appreciated.

Re-testing tier1-6 and stress testing.

**Review by TobiHartmann — APPROVED** [review]
Still looks good to me.

**Comment by eme64** [comment]
I had to add an `avx2` requirement to the test. The unaligned example crashed otherwise, see `Test21.java` [here](https://bugs.openjdk.org/browse/JDK-8310190?focusedCommentId=14615859&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14615859).

It creates this assert, because the unsafe int memory access is at offset 6, which is not aligned. And somehow, on `-XX:UseAVX=0 -XX:UseSSE=3` we require alignment.
`assert((ABS(iv_adjustment_in_bytes) % elt_size) == 0 || !vectors_should_be_aligned()) failed: (6) should be divisible by (4)`

**Review by vnkozlov — APPROVED** [review]
Update looks good.

**Comment by eme64** [comment]
Thanks @vnkozlov @TobiHartmann for the reviews!
/integrate

**Pull request closed without merging**